File: adapters/obsidian/src/stigmem_obsidian/parser.py
"""Markdown note parser — frontmatter, wikilinks, and Dataview inline fields."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


# ---------------------------------------------------------------------------
# Data shapes
# ---------------------------------------------------------------------------

@dataclass
class ParsedNote:
    """Structured representation of a parsed markdown note."""

    rel_path: str
    title: str
    frontmatter: dict[str, Any]
    wikilinks: list[str]
    dataview_fields: dict[str, str]
    body: str
    content_hash: str

    @property
    def entity_uri(self) -> str:
        """Stable entity URI derived from vault-relative path (no extension)."""
        stem = self.rel_path.removesuffix(".md")
        return f"obsidian://vault/{stem}"


# ---------------------------------------------------------------------------
# Regex patterns
# ---------------------------------------------------------------------------

# [[wikilink]] or [[wikilink|alias]]
_WIKILINK_RE = re.compile(r"\[\[([^\]|#]+?)(?:[|#][^\]]*?)?\]\]")

# Dataview inline field: key:: value (anywhere on line, not in code blocks)
_DATAVIEW_RE = re.compile(r"^([A-Za-z][A-Za-z0-9_ -]*)::[ \t]*(.+)$", re.MULTILINE)

# The managed Stigmem section header
STIGMEM_SECTION_HEADER = "## Stigmem"
STIGMEM_SECTION_RE = re.compile(
    r"(^## Stigmem\n)(.*?)(?=^## |\Z)",
    re.MULTILINE | re.DOTALL,
)


# ---------------------------------------------------------------------------
# Parser
# ---------------------------------------------------------------------------

def parse_note(path: Path, vault_root: Path) -> ParsedNote:
    """Parse a single markdown note file into a ParsedNote.

    Handles YAML frontmatter (via python-frontmatter), wikilinks, and
    Dataview inline fields. Skips code-block content for Dataview extraction.
    """
    import frontmatter as fm
    import hashlib

    raw_text = path.read_text(encoding="utf-8", errors="replace")
    post = fm.loads(raw_text)

    meta: dict[str, Any] = dict(post.metadata)
    body: str = post.content

    title = str(meta.get("title", "") or path.stem)

    wikilinks = _extract_wikilinks(body)

    body_no_code = _strip_code_blocks(body)
    dataview_fields = _extract_dataview(body_no_code)

    content_hash = hashlib.sha256(raw_text.encode()).hexdigest()[:16]

    rel_path = str(path.relative_to(vault_root)).replace("\\", "/")

    return ParsedNote(
        rel_path=rel_path,
        title=title,
        frontmatter=meta,
        wikilinks=wikilinks,
        dataview_fields=dataview_fields,
        body=body,
        content_hash=content_hash,
    )


def _extract_wikilinks(text: str) -> list[str]:
    """Return deduplicated list of wikilink targets from text."""
    seen: set[str] = set()
    result: list[str] = []
    for m in _WIKILINK_RE.finditer(text):
        target = m.group(1).strip()
        if target and target not in seen:
            seen.add(target)
            result.append(target)
    return result


def _extract_dataview(text: str) -> dict[str, str]:
    """Extract Dataview inline fields (key:: value) from text."""
    return {m.group(1).strip(): m.group(2).strip() for m in _DATAVIEW_RE.finditer(text)}


def _strip_code_blocks(text: str) -> str:
    """Remove fenced code blocks so Dataview regex doesn't match inside them."""
    return re.sub(r"```.*?```", "", text, flags=re.DOTALL)


# ---------------------------------------------------------------------------
# Stigmem-section helpers
# ---------------------------------------------------------------------------

def extract_stigmem_section(content: str) -> str | None:
    """Return the body of the `## Stigmem` section, or None if absent."""
    m = STIGMEM_SECTION_RE.search(content)
    if m:
        return m.group(2).strip()
    return None


def replace_stigmem_section(content: str, new_body: str) -> str:
    """Replace the `## Stigmem` section body, or append the section if absent."""
    section_text = f"{STIGMEM_SECTION_HEADER}\n{new_body}\n"
    if STIGMEM_SECTION_RE.search(content):
        return STIGMEM_SECTION_RE.sub(
            lambda m: section_text,
            content,
            count=1,
        )
    sep = "\n" if content.endswith("\n") else "\n\n"
    return content + sep + section_text


def build_stigmem_section_body(facts: list[dict[str, Any]]) -> str:
    """Render a list of fact dicts as a YAML-ish block for the Stigmem section."""
    lines: list[str] = []
    for fact in facts:
        relation = fact.get("relation", "")
        value = fact.get("value", "")
        source = fact.get("source", "")
        lines.append(f"- relation: {relation}")
        lines.append(f"  value: {value}")
        if source:
            lines.append(f"  source: {source}")
    return "\n".join(lines)


def parse_stigmem_section_body(body: str) -> list[dict[str, str]]:
    """Parse the YAML-ish Stigmem section body back into a list of fact dicts."""
    facts: list[dict[str, str]] = []
    current: dict[str, str] = {}
    for line in body.splitlines():
        stripped = line.strip()
        if stripped.startswith("- relation:"):
            if current:
                facts.append(current)
            current = {"relation": stripped[len("- relation:"):].strip()}
        elif stripped.startswith("value:") and current:
            current["value"] = stripped[len("value:"):].strip()
        elif stripped.startswith("source:") and current:
            current["source"] = stripped[len("source:"):].strip()
    if current:
        facts.append(current)
    return facts


def add_conflict_comment(content: str, conflict_note: str) -> str:
    """Append an Obsidian markdown comment surfacing a conflict."""
    comment = f"\n%%stigmem-conflict: {conflict_note}%%\n"
    return content + comment

File: adapters/obsidian-plugin/src/parser.ts
/**
 * Markdown note parser — frontmatter, wikilinks, Dataview inline fields, and the managed
 * ## Stigmem section. Mirrors the Python adapter's parser.py semantics exactly.
 */

// ---------------------------------------------------------------------------
// Regex patterns (mirrors Python adapter)
// ---------------------------------------------------------------------------

// [[wikilink]] or [[wikilink|alias]] or [[wikilink#heading]]
const WIKILINK_RE = /\[\[([^\]|#]+?)(?:[|#][^\]]*?)?\]\]/g;

// Dataview inline field: key:: value
const DATAVIEW_RE = /^([A-Za-z][A-Za-z0-9_ -]*)::[ \t]*(.+)$/gm;

// Fenced code blocks to strip before Dataview extraction
const CODE_BLOCK_RE = /```[\s\S]*?```/g;

export const STIGMEM_SECTION_HEADER = "## Stigmem";
const STIGMEM_SECTION_RE = /^## Stigmem\n([\s\S]*?)(?=^## |\Z)/m;

// ---------------------------------------------------------------------------
// Data shapes
// ---------------------------------------------------------------------------

export interface ParsedNote {
	relPath: string;
	title: string;
	frontmatter: Record<string, unknown>;
	wikilinks: string[];
	dataviewFields: Record<string, string>;
	body: string;
	contentHash: string;
	entityUri: string;
}

export interface StigmemFact {
	relation: string;
	value: string;
	source?: string;
}

// ---------------------------------------------------------------------------
// YAML frontmatter parsing (no external deps — inline minimal parser)
// ---------------------------------------------------------------------------

function parseFrontmatter(content: string): { meta: Record<string, unknown>; body: string } {
	if (!content.startsWith("---")) {
		return { meta: {}, body: content };
	}
	const end = content.indexOf("\n---", 4);
	if (end === -1) {
		return { meta: {}, body: content };
	}
	const yamlBlock = content.slice(4, end);
	const body = content.slice(end + 4).replace(/^\n/, "");
	const meta = parseYamlBlock(yamlBlock);
	return { meta, body };
}

/** Minimal YAML parser for simple scalar/list frontmatter (no nested maps). */
function parseYamlBlock(yaml: string): Record<string, unknown> {
	const result: Record<string, unknown> = {};
	const lines = yaml.split("\n");
	let i = 0;
	while (i < lines.length) {
		const line = lines[i];
		const colonIdx = line.indexOf(":");
		if (colonIdx === -1) { i++; continue; }
		const key = line.slice(0, colonIdx).trim();
		const rest = line.slice(colonIdx + 1).trim();
		if (!key) { i++; continue; }

		if (rest === "" || rest === "|" || rest === ">") {
			// Check if next lines are a YAML list
			const items: string[] = [];
			i++;
			while (i < lines.length && lines[i].match(/^[ \t]+-\s+/)) {
				items.push(lines[i].replace(/^[ \t]+-\s+/, "").trim());
				i++;
			}
			if (items.length > 0) {
				result[key] = items;
			} else {
				result[key] = null;
			}
			continue;
		}

		// Inline list: [a, b, c]
		if (rest.startsWith("[") && rest.endsWith("]")) {
			result[key] = rest.slice(1, -1).split(",").map(s => s.trim()).filter(Boolean);
			i++;
			continue;
		}

		// Quoted string
		if ((rest.startsWith('"') && rest.endsWith('"')) ||
			(rest.startsWith("'") && rest.endsWith("'"))) {
			result[key] = rest.slice(1, -1);
			i++;
			continue;
		}

		// Boolean
		if (rest === "true") { result[key] = true; i++; continue; }
		if (rest === "false") { result[key] = false; i++; continue; }

		// Number
		const num = Number(rest);
		if (!isNaN(num) && rest !== "") { result[key] = num; i++; continue; }

		result[key] = rest;
		i++;
	}
	return result;
}

// ---------------------------------------------------------------------------
// Content hash (FNV-32a — cheap, no crypto API needed in the browser)
// ---------------------------------------------------------------------------

function fnv32a(text: string): string {
	let hash = 0x811c9dc5;
	for (let i = 0; i < text.length; i++) {
		hash ^= text.charCodeAt(i);
		hash = (hash * 0x01000193) >>> 0;
	}
	return hash.toString(16).padStart(8, "0");
}

// ---------------------------------------------------------------------------
// Public parsers
// ---------------------------------------------------------------------------

export function parseNote(relPath: string, content: string): ParsedNote {
	const { meta, body } = parseFrontmatter(content);

	const title = (meta["title"] as string | undefined) ?? relPath.replace(/\.md$/, "").split("/").pop() ?? relPath;

	const wikilinks = extractWikilinks(body);
	const bodyNoCode = body.replace(CODE_BLOCK_RE, "");
	const dataviewFields = extractDataview(bodyNoCode);
	const contentHash = fnv32a(content);

	const stem = relPath.replace(/\.md$/, "");
	const entityUri = `obsidian://vault/${stem}`;

	return { relPath, title, frontmatter: meta, wikilinks, dataviewFields, body, contentHash, entityUri };
}

function extractWikilinks(text: string): string[] {
	const seen = new Set<string>();
	const result: string[] = [];
	let match: RegExpExecArray | null;
	const re = new RegExp(WIKILINK_RE.source, WIKILINK_RE.flags);
	while ((match = re.exec(text)) !== null) {
		const target = match[1].trim();
		if (target && !seen.has(target)) {
			seen.add(target);
			result.push(target);
		}
	}
	return result;
}

function extractDataview(text: string): Record<string, string> {
	const result: Record<string, string> = {};
	const re = new RegExp(DATAVIEW_RE.source, DATAVIEW_RE.flags);
	let match: RegExpExecArray | null;
	while ((match = re.exec(text)) !== null) {
		result[match[1].trim()] = match[2].trim();
	}
	return result;
}

// ---------------------------------------------------------------------------
// Stigmem section helpers (mirrors Python parser.py)
// ---------------------------------------------------------------------------

export function extractStigmemSection(content: string): string | null {
	const m = STIGMEM_SECTION_RE.exec(content);
	return m ? m[1].trim() : null;
}

export function replaceStigmemSection(content: string, newBody: string): string {
	const sectionText = `${STIGMEM_SECTION_HEADER}\n${newBody}\n`;
	if (STIGMEM_SECTION_RE.test(content)) {
		return content.replace(STIGMEM_SECTION_RE, sectionText);
	}
	const sep = content.endsWith("\n") ? "\n" : "\n\n";
	return content + sep + sectionText;
}

export function buildStigmemSectionBody(facts: StigmemFact[]): string {
	const lines: string[] = [];
	for (const fact of facts) {
		lines.push(`- relation: ${fact.relation}`);
		lines.push(`  value: ${fact.value}`);
		if (fact.source) lines.push(`  source: ${fact.source}`);
	}
	return lines.join("\n");
}

export function parseStigmemSectionBody(body: string): StigmemFact[] {
	const facts: StigmemFact[] = [];
	let current: Partial<StigmemFact> = {};
	for (const line of body.split("\n")) {
		const s = line.trim();
		if (s.startsWith("- relation:")) {
			if (current.relation) facts.push(current as StigmemFact);
			current = { relation: s.slice("- relation:".length).trim() };
		} else if (s.startsWith("value:") && current.relation) {
			current.value = s.slice("value:".length).trim();
		} else if (s.startsWith("source:") && current.relation) {
			current.source = s.slice("source:".length).trim();
		}
	}
	if (current.relation) facts.push(current as StigmemFact);
	return facts;
}

export function addConflictComment(content: string, conflictNote: string): string {
	return content + `\n%%stigmem-conflict: ${conflictNote}%%\n`;
}
